lib/publiccloud/ec2.pm

```diff
@@ -17,13 +17,60 @@ use testapi;
 
 has ssh_key      => undef;
 has ssh_key_file => undef;
+has credentials  => undef;
+
+sub create_credentials {
+    my ($self) = @_;
+
+    if (!defined($self->credentials)) {
+        my $credentials = $self->do_rest_api('/ec2/users', method => 'post');
+        die("Missing key") if (!defined($credentials->{keys}));
+        for my $key (@{$credentials->{keys}}) {
+            next if (!defined($key->{secret}));
+            next if ($key->{status} ne 'active');
+            $self->key_id($key->{key_id});
+            $self->key_secret($key->{secret});
+        }
+        $self->credentials($credentials);
+        die('Failed to create credentials') unless (defined($self->key_id) && defined($self->key_secret));
+    }
+
+    return $self->credentials;
+}
+
+sub delete_credentials {
+    my ($self) = @_;
+
+    return unless (defined($self->credentials));
+
+    my $username = $self->credentials->{name};
+    $self->do_rest_api('/credentials/users/' . $username, method => 'delete');
+    $self->credentials(undef);
+}
+
+sub _check_credentials {
+    my ($self) = @_;
+    my $max_tries = 6;
+    for my $i (1 .. $max_tries) {
+        my $out = script_output('aws ec2 describe-images --dry-run', 60, proceed_on_failure => 1);
+        return 1 if ($out !~ /AuthFailure/m);
+        sleep 30;
+    }
+}
 
 sub init {
     my ($self, %params) = @_;
     $self->SUPER::init();
+
+    if (!defined($self->key_id) || !defined($self->key_secret)) {
```

Comment:
I think we need AND here. if we will have for some reason only key or only secret it won't fly ...

```suggestion
   $self->create_credentials() unless $self->key_id && $self->key_secret;
```

or to keep following "defined"  approach :

```suggestion
     if (!(defined($self->key_id) && defined($self->key_secret))) {
```

Reply:
Isn't it the same?

Reply:
I am wrong here , please ignore